src/test/java/io/quarkus/gizmo/ClassTransformerTest.java

```diff
@@ -0,0 +1,105 @@
+        cv.visitSource(null, null);
+
+        // Add constructor
+        MethodVisitor mv = cv.visitMethod(ACC_PUBLIC, MethodDescriptor.INIT, "()V", null, null);
+        mv.visitVarInsn(ALOAD, 0); // push `this` to the operand stack
+        mv.visitMethodInsn(INVOKESPECIAL, objectName, MethodDescriptor.INIT, "()V", false); // call the constructor of super class
+        mv.visitInsn(RETURN);
+        mv.visitMaxs(0, 1);
+        mv.visitEnd();
+
+        ClassTransformer classTransformer = new ClassTransformer(cv);
+
+        // private String latVal;
+        FieldCreator lastValue = classTransformer.addField(FieldDescriptor.of(className, "lastVal", String.class));
+        lastValue.setModifiers(ACC_PRIVATE);
+
+        // public final String transform(String val) {
+        //    if (lastVal != null && lastVal.equals("bar")) {
+        //       return "baZ";
+        //    }
+        //    lastVal = val;
+        //    if (val.equals("bar")) {
+        //       return val;
+        //    }
+        //    return val.toUpperCase();
+        // }
+        MethodCreator transform = classTransformer
+                .addMethod(MethodDescriptor.ofMethod(className, "transform", String.class, String.class));
+        transform.setModifiers(ACC_PUBLIC | ACC_FINAL);
+
+        ResultHandle lastVal = transform.readInstanceField(lastValue.getFieldDescriptor(), transform.getThis());
+        BytecodeCreator isLastValNotNull = transform.ifNotNull(lastVal).trueBranch();
+        BytecodeCreator isLastValBar = isLastValNotNull
+                .ifTrue(Gizmo.equals(isLastValNotNull, lastVal, isLastValNotNull.load("bar")))
+                .trueBranch();
+        isLastValBar.returnValue(isLastValBar.load("baZ"));
+
+        transform.writeInstanceField(lastValue.getFieldDescriptor(), transform.getThis(), transform.getMethodParam(0));
+
+        BytecodeCreator isBar = transform.ifTrue(Gizmo.equals(transform, transform.getMethodParam(0), transform.load("bar")))
+                .trueBranch();
+        isBar.returnValue(transform.getMethodParam(0));
+
+        ResultHandle ret = transform.invokeVirtualMethod(MethodDescriptor.ofMethod(String.class, "toUpperCase", String.class),
+                transform.getMethodParam(0));
+
+        assertThrows(UnsupportedOperationException.class, () -> transform.createFunction(Function.class));
+        assertThrows(UnsupportedOperationException.class, () -> transform.enumSwitch(null, null));
+
+        transform.returnValue(ret);
+
+        classTransformer.close();
+
+        cv.visitEnd();
+        cl.write(className, cv.toByteArray());
+
+        Class<?> clazz = cl.loadClass("com.MyTest");
+        assertTrue(clazz.isSynthetic());
+        MyInterface myInterface = (MyInterface) clazz.getDeclaredConstructor().newInstance();
+        Method transformMethod = clazz.getMethod("transform", String.class);
+        assertTrue(Modifier.isFinal(transformMethod.getModifiers()));
```

Comment:
I would use try-with-resources for the `ClassTransformer` and use reflection here to also assert something about the field.

Reply:
I don't think that we need to test the `AutoCloseable` functionality... Field assertion added.